crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -30,6 +30,194 @@ use multimap::MultiMap;
+    Xoring,
+    PopStack1ReplacingWithCheck(&'names str),
+    Simple,
+}
+
+/// This machine is used to check that there are no instructions after next
+/// in a fold block over stream and map, e.g.
+/// (fold $iterable iterator
+///  (seq
+///    (next iterator)
+///    (call  ...)  <- instruction after next
+///  )
+/// )
+/// Note that the fold over scalar doesn't have this restriction.
+#[derive(Clone, Debug)]
+struct AfterNextCheckMachine<'name> {
+    /// stack for the machine.
+    stack: Vec<(CheckInstructionKind<'name>, Span)>,
+
+    /// lalrpop effectively parses from right to left thus `next`
+    /// happens before fold. It is impossible to tell whether `next`
+    /// belongs to a scalar fold or not.
+    potentially_impossible_spans: Vec<(&'name str, Span)>,
+
+    /// This vector contains all spans where an instruction after next was met.
+    impossible_spans: Vec<Span>,
+
+    /// This flag disables the machine if invariants are broken, e.g. par/seq must see
+    /// at least 2 instruction kinds in the stack.
+    is_enabled: bool,
+}
+
+impl<'name> Default for AfterNextCheckMachine<'name> {
+    fn default() -> Self {
+        Self {
+            stack: Default::default(),
+            potentially_impossible_spans: Default::default(),
+            impossible_spans: Default::default(),
+            is_enabled: true,
+        }
+    }
+}
+
+impl<'name> AfterNextCheckMachine<'name> {
+    fn impossible_spans_mut_iter(&mut self) -> std::slice::IterMut<Span> {
+        self.impossible_spans.iter_mut()
+    }
+
+    fn met_instruction_kind(&mut self, instr_kind: CheckInstructionKind<'name>, span: Span) {
+        if !self.is_enabled {
+            return;
+        }
+        match instr_kind {
+            CheckInstructionKind::Replacing => {
+                let child = self.stack.pop();
+                match child {
+                    Some((CheckInstructionKind::PivotalNext(name), _)) => {
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(name), span));
+                    }
+                    Some(..) => {
+                        self.stack.push((CheckInstructionKind::Replacing, span));
+                    }
+                    None => self.is_enabled = false,
+                }
+            }
+            CheckInstructionKind::Xoring => {
+                let right_branch = self.stack.pop();
+                let left_branch = self.stack.pop();
+                match (left_branch, right_branch) {
+                    (
+                        Some((CheckInstructionKind::PivotalNext(..), _)),
+                        Some((CheckInstructionKind::PivotalNext(..), _)),
+                    ) => {
+                        // `xor` has `next` in both branches. It is impossible
+                        // to tell which branch should poped up.
+                        self.stack.clear();
+                        self.is_enabled = false;
```

Comment:
Let's introduce `self.disable()` given the statements above are repeated several times.

Reply:
Done